File: apps/public-api/docs/webhooks/verifiers/verify.py
"""
Reference verifier for Contractor Ops outbound webhooks (Python 3, stdlib only).

Each delivery carries `X-CO-Signature: t={unix_ms},v1={hex}` where
    v1 = HMAC_SHA256(secret, f"{t}.{raw_body}")  (hex).

Verify with the per-subscription secret shown once at creation. Reject any
signature whose timestamp is outside a 5-minute window BEFORE comparing the
digest (this is your replay protection), then compare in constant time.
"""

import hashlib
import hmac
import time

FIVE_MINUTES_MS = 5 * 60 * 1000


def verify_webhook_signature(
    secret: str,
    raw_body: bytes,
    signature_header: str,
    now_ms: int | None = None,
    tolerance_ms: int = FIVE_MINUTES_MS,
) -> bool:
    if now_ms is None:
        now_ms = int(time.time() * 1000)

    t: int | None = None
    v1: str | None = None
    for part in signature_header.split(","):
        key, _, value = part.strip().partition("=")
        if key == "t":
            try:
                t = int(value)
            except ValueError:
                return False
        elif key == "v1":
            v1 = value.lower()

    if t is None or v1 is None or len(v1) != 64:
        return False

    # Replay window — reject stale/future timestamps BEFORE the digest compare.
    if abs(now_ms - t) > tolerance_ms:
        return False

    signed = f"{t}.".encode("utf-8") + raw_body
    expected = hmac.new(secret.encode("utf-8"), signed, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, v1)

File: apps/public-api/docs/webhooks/verifiers/verify.ts
/**
 * Reference verifier for Contractor Ops outbound webhooks (TypeScript / Node.js).
 *
 * Each delivery carries `X-CO-Signature: t={unix_ms},v1={hex}` where
 *   v1 = HMAC_SHA256(secret, `${t}.${rawBody}`)  (hex).
 *
 * Verify with the per-subscription secret shown once at creation. Reject any
 * signature whose timestamp is outside a 5-minute window BEFORE comparing the
 * digest — this is your replay protection. Compare in constant time.
 *
 * Dependency-free: Node.js stdlib `crypto` only.
 */

import { createHmac, timingSafeEqual } from 'node:crypto';

const FIVE_MINUTES_MS = 5 * 60 * 1000;

export function verifyWebhookSignature(
  secret: string,
  rawBody: string,
  signatureHeader: string,
  nowMs: number = Date.now(),
  toleranceMs: number = FIVE_MINUTES_MS,
): boolean {
  let t: number | undefined;
  let v1: string | undefined;
  for (const part of signatureHeader.split(',')) {
    const eq = part.indexOf('=');
    if (eq === -1) continue;
    const key = part.slice(0, eq).trim();
    const value = part.slice(eq + 1).trim();
    if (key === 't') t = Number(value);
    else if (key === 'v1') v1 = value;
  }
  if (t === undefined || v1 === undefined || !Number.isFinite(t) || !/^[0-9a-f]{64}$/i.test(v1)) {
    return false;
  }

  // Replay window — reject stale/future timestamps BEFORE the digest compare.
  if (Math.abs(nowMs - t) > toleranceMs) return false;

  const expected = createHmac('sha256', secret).update(`${t}.${rawBody}`).digest('hex');
  const a = Buffer.from(v1, 'hex');
  const b = Buffer.from(expected, 'hex');
  if (a.length !== b.length) return false;
  return timingSafeEqual(a, b);
}
